packages/babel-traverse/src/path/lib/virtual-types-validator.ts

```diff
@@ -37,81 +37,57 @@ type Opts<Obj> = Partial<{
-    opts?: VirtualTypeAliases["NumericLiteralTypeAnnotation"],
-  ): void;
-  isPure(opts?: VirtualTypeAliases["Pure"]): boolean;
-  isReferenced(opts?: VirtualTypeAliases["Referenced"]): boolean;
-  isReferencedIdentifier(
+  isNumericLiteralTypeAnnotation(this: NodePath): void;
+  isPure(): boolean;
+  isReferenced(): boolean;
+  isReferencedIdentifier<
+    Opts extends Options<VirtualTypeAliases["ReferencedIdentifier"]>,
+  >(
     this: NodePath,
-    opts?: Opts<VirtualTypeAliases["ReferencedIdentifier"]>,
-  ): this is NodePath<VirtualTypeAliases["ReferencedIdentifier"]>;
+    opts?: Opts,
+  ): this is NodePath<VirtualTypeAliases["ReferencedIdentifier"] & Opts>;
   isReferencedMemberExpression(
     this: NodePath,
-    opts?: Opts<VirtualTypeAliases["ReferencedMemberExpression"]>,
   ): this is NodePath<VirtualTypeAliases["ReferencedMemberExpression"]>;
-  isRestProperty(
-    this: NodePath,
-    opts?: Opts<VirtualTypeAliases["RestProperty"]>,
-  ): this is NodePath<t.RestProperty>;
-  isScope(
-    this: NodePath,
-    opts?: Opts<VirtualTypeAliases["Scope"]>,
-  ): this is NodePath<VirtualTypeAliases["Scope"]>;
-  isSpreadProperty(
-    this: NodePath,
-    opts?: Opts<VirtualTypeAliases["SpreadProperty"]>,
-  ): this is NodePath<t.SpreadProperty>;
-  isStatement(
-    this: NodePath,
-    opts?: Opts<VirtualTypeAliases["Statement"]>,
-  ): this is NodePath<t.Statement>;
-  isUser(opts?: VirtualTypeAliases["User"]): boolean;
-  isVar(
-    this: NodePath,
-    opts?: Opts<VirtualTypeAliases["Var"]>,
-  ): this is NodePath<VirtualTypeAliases["Var"]>;
-}
-
-export function isReferencedIdentifier(this: NodePath, opts?: any): boolean {
+  isRestProperty(this: NodePath): this is NodePath<t.RestProperty>;
+  isScope(this: NodePath): this is NodePath<VirtualTypeAliases["Scope"]>;
+  isSpreadProperty(this: NodePath): this is NodePath<t.SpreadProperty>;
+  isStatement(this: NodePath): this is NodePath<t.Statement>;
+  isUser(): this is NodePath<VirtualTypeAliases["User"]>;
+  isVar(this: NodePath): this is NodePath<VirtualTypeAliases["Var"]>;
+}
+
+export function isReferencedIdentifier(
+  this: NodePath,
+  opts?: Options<VirtualTypeAliases["ReferencedIdentifier"]>,
+): boolean {
   const { node, parent } = this;
-  if (!isIdentifier(node, opts) && !isJSXMemberExpression(parent, opts)) {
-    if (isJSXIdentifier(node, opts)) {
-      if (isCompatTag(node.name)) return false;
-    } else {
-      // not a JSXIdentifier or an Identifier
-      return false;
-    }
+  if (isIdentifier(node, opts)) {
+    return nodeIsReferenced(node, parent, this.parentPath.parent);
+  } else if (isJSXIdentifier(node, opts)) {
+    if (!isJSXMemberExpression(parent) && isCompatTag(node.name)) return false;
+    return nodeIsReferenced(node, parent, this.parentPath.parent);
+  } else {
+    // not a JSXIdentifier or an Identifier
+    return false;
   }
-
-  // check if node is referenced
-  return nodeIsReferenced(node, parent, this.parentPath.parent);
```

Comment:
Previously this branch also handles the JSXIdentifier in `JSXMemberExpression` case, however at this point we  the `opts` are no longer checked against the JSXIdentifier node, so `NodePath(<A.B/>).get(path to A).isReferencedIdentifier({ name: "NotA" })` would incorrectly return `true`. This is also fixed in the new implementation.

Reply:
Could you update the PR title so that it properly describes this in the changelog?